addons/account/models/account_tax.py

```diff
@@ -664,6 +665,418 @@ def recompute_base(base_amount, fixed_amount, percent_amount, division_amount):
             'total_void': sign * currency.round(total_void),
         }
 
+    @api.model
+    def _get_line_field_value(self, line, field_name, value, relational_model=None, default_value=None):
+        if value is not None:
+            return value
+        if field_name in line._fields:
+            return line[field_name]
+        return self.env[relational_model] if relational_model else default_value
+
+    @api.model
+    def _prepare_base_line(self, base_line,
+                           partner=None, currency=None, product=None, taxes=None, price_unit=None, quantity=None,
+                           discount=None, account=None, analytic_account=None, analytic_tags=None, price_subtotal=None,
+                           is_refund=False):
+        return {
+            'record': base_line,
+            'partner': self._get_line_field_value(
+                base_line, 'partner_id', partner, relational_model='res.partner'),
+            'currency': self._get_line_field_value(
+                base_line, 'currency_id', currency, relational_model='res.currency'),
+            'product': self._get_line_field_value(
+                base_line, 'product_id', product, relational_model='product.product'),
+            'taxes': self._get_line_field_value(
+                base_line, 'tax_ids', taxes, relational_model='account.tax'),
+            'price_unit': self._get_line_field_value(
+                base_line, 'price_unit', price_unit, default_value=0.0),
+            'quantity': self._get_line_field_value(
+                base_line, 'quantity', quantity, default_value=1.0),
+            'discount': self._get_line_field_value(
+                base_line, 'discount', discount, default_value=0.0),
+            'account': self._get_line_field_value(
+                base_line, 'account_id', account, relational_model='account.account'),
+            'analytic_account': self._get_line_field_value(
+                base_line, 'analytic_account_id', analytic_account, relational_model='account.analytic.account'),
+            'analytic_tags': self._get_line_field_value(
+                base_line, 'analytic_tag_ids', analytic_tags, relational_model='account.analytic.tag'),
+            'price_subtotal': self._get_line_field_value(
+                base_line, 'price_subtotal', price_subtotal, default_value=0.0),
+            'is_refund': is_refund,
+        }
```

Comment:
Please stop using this magic fallback logic.
Ok for a default fallback to `1.0|0.0|empty recordset` for values not provided, but you shouldn't fallback magically on fields from other models.

When the method is called, all the available parameters must be given, without relying on some magic defaults fetched from the record fields.

Otherwise we don't know what's used for the tax computation, where it comes from clearly, ...

Reply:
I agree with you this is not what we do in odoo but if I add a method on each business model converting to this dictionnary, I don't get the advantage to not use my generic mixin...
I'm trying to make this the most easy to use as possible. If you don't like it, I'll change this part then.

Reply:
Can't you specify all the needed parameters when this method is called ?
You don't have to add a method on each model, only to give more parameters to the `_prepare_base_line` calls